### PR TITLE
Part 3: Add dictionary based counter

### PR DESCRIPTION
**Pull request by benjimarshall**
Part 2 was to use regex, which I used as when doing part 1 so I moved straight onto part 3.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -9,16 +9,34 @@ function readSampleText(): string {
...
     // treated like individual characters in the trigram
-    return input.substring(offset).match(/.{1,3}/gs) ?? [];
+    return input.substring(offset).match(/.{3}/gs) ?? [];
 }
 
 function getAllTrigrams(input: string): string[] {
     // Make list of all trigrams in input by splitting into chunks at offsets 0, 1, and 2
-    return _.flatMap(_.range(0, 3), i => getTrigramsFromOffset(input, i));
+    return _.flatMap(_.range(0, 3), i => getTrigramsFromOffset(input, i))
+        .filter(s => /[a-z]{3}/.test(s));
```
Comment (FridaTveit) [inline src/index.ts:18]: What's this filter for? 🙂

File: src/index.ts [hunk 2]
```diff
@@ -9,16 +9,34 @@ function readSampleText(): string {
...
 }
 
-console.log(naiveTls(readSampleText()));
+function dictionaryTls(input: string): _.Dictionary<number> {
+    const trigrams = getAllTrigrams(input.toLowerCase());
+
+    const dictionary: _.Dictionary<number> = {};
+    trigrams.map(trigram =>
+        dictionary[trigram] = (dictionary[trigram] ?? 0) + 1
+    );
```
Comment (FridaTveit) [inline src/index.ts:31]: [countBy](https://lodash.com/docs/4.17.15#countBy) should be able to do this for you I believe, giving you more functional code 🙂

File: src/index.ts [hunk 3]
```diff
@@ -9,16 +9,34 @@ function readSampleText(): string {
...
+
+    return dictionary;
+}
+
+function getEntriesWithValue(dictionary: _.Dictionary<number>, target: number): string[] {
+    return Object.entries(dictionary)
```
Comment (FridaTveit) [inline src/index.ts:37]: You should be able to just use [filter](https://lodash.com/docs/4.17.15#filter) here 🙂

File: src/index.ts [hunk 4]
```diff
@@ -9,16 +9,34 @@ function readSampleText(): string {
...
+
+    const dictionary: _.Dictionary<number> = {};
+    trigrams.map(trigram =>
+        dictionary[trigram] = (dictionary[trigram] ?? 0) + 1
+    );
+
+    return dictionary;
+}
+
+function getEntriesWithValue(dictionary: _.Dictionary<number>, target: number): string[] {
```
Comment (FridaTveit) [inline src/index.ts:36]: For the dictionary I would expect you to use the type:
```
{ [trigram: string]: number; }
```
🙂

**Review by FridaTveit — COMMENTED** [review]
A few comments, otherwise looks good 🙂

File: src/index.ts (hunk 1 above)
Comment (benjimarshall) [reply src/index.ts]: This filter ensures that the trigrams are three letter strings, and discards trigrams containing whitespace or punctuation, as per the Three Letter String definition we were given. I've refactored this step out into its own function in c2ed0d5 to make this a bit clearer 🙂

File: src/index.ts (hunk 2 above)
Comment (benjimarshall) [reply src/index.ts]: Ah that's exactly what I need! Changed to it in c2ed0d5 🙂

File: src/index.ts (hunk 3 above)
Comment (benjimarshall) [reply src/index.ts]: I ended up using the [`_.pickBy`](lodash.com/docs/4.17.15#pickBy) function in c2ed0d5 to tidy this up, it's a version of `filter` that filters on the values of items in an object. I hope that's ok 🙂

File: src/index.ts (hunk 4 above)
Comment (benjimarshall) [reply src/index.ts]: Yeah that's much tidier, thank you! Changed in c2ed0d5 🙂

File: src/index.ts [hunk 5]
```diff
@@ -14,29 +14,26 @@ function getTrigramsFromOffset(input: string, offset: number): string[] {
...
 
-function getEntriesWithValue(dictionary: _.Dictionary<number>, target: number): string[] {
-    return Object.entries(dictionary)
-        .filter(([, value]) => value === target)
-        .map(([key]) => key);
+function getEntriesWithValue(
+    dictionary: { [trigram: string]: number; },
+    target: number
+): string[] {
+    return _.keys(_.pickBy(dictionary, x => x === (target)));
```
Comment (FridaTveit) [inline src/index.ts:36]: Might be able to slightly simplify by:
```
_(dictionary)
  .pickBy(x => x === target)
  .keys();
```
and don't need brackets around target 🙂

**Review by FridaTveit — COMMENTED** [review]
One minor suggestion, otherwise looks great 🙂

File: src/index.ts (hunk 5 above)
Comment (benjimarshall) [reply src/index.ts]: That looks cleaner, thank you! Changed in 53d57aa 🙂

**Review by FridaTveit — APPROVED** [review]
Looks great 🙂

**Pull request merged**